src/Data/Poly/Internal/Dense/Laurent.hs

```diff
@@ -0,0 +1,284 @@
+-- |
+-- Module:     Data.Poly.Internal.Laurent
+-- Copyright:  (c) 2019 Adjoint Inc
+-- Licence:    BSD3
+-- Maintainer: Adjoint Inc <info@adjoint.io>
+--
+-- Laurent polynomials of one variable.
+--
+{-# LANGUAGE DeriveAnyClass      #-}
+{-# LANGUAGE DeriveGeneric       #-}
+{-# LANGUAGE DerivingStrategies  #-}
+{-# LANGUAGE FlexibleContexts    #-}
+{-# LANGUAGE PatternSynonyms     #-}
+{-# LANGUAGE RankNTypes          #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE ViewPatterns        #-}
+
+
+module Data.Poly.Internal.Dense.Laurent
+  ( Laurent(..)
+  , ULaurent
+  , VLaurent
+  , leading
+  , (^-)
+  -- * Num interface
+  , toLaurent
+  , monomial
+  , scale
+  , pattern X
+  , eval
+  , subst
+  -- , deriv
+  -- , integral
+
+   -- * Semiring interface
+  , toLaurent'
+  , monomial'
+  -- XXX: Add remaining
+  )
+where
+
+import Control.DeepSeq             (NFData)
+import Control.Monad
+import Control.Monad.ST
+import Data.List (intersperse)
+import Data.Poly.Internal.Dense (Poly (..), toPoly, toPoly')
+import qualified Data.Poly.Internal.Dense    as Poly
+import Data.Semiring (Semiring (..))
+import qualified Data.Semiring as S
+import qualified Data.Vector as V
+import qualified Data.Vector.Generic as G
+import qualified Data.Vector.Generic.Mutable as MG
+import qualified Data.Vector.Unboxed as U
+import GHC.Generics (Generic)
+
+-- TODO: Move this to Internal/Dense/Laurent.hs
+
+data Laurent v a = Laurent !Int !(Poly v a)
+  deriving (Eq, Ord, Generic)
+  deriving anyclass (NFData)
+
+type VLaurent a = Laurent V.Vector a
+type ULaurent a = Laurent U.Vector a
+
+instance (Eq a, Num a, G.Vector v a) => Num (Laurent v a) where
+  Laurent n xs * Laurent m ys = Laurent (n + m) (xs * ys)
+  Laurent n xs + Laurent m ys
+    | n <= m    = Laurent n (xs + Poly.scale (fromIntegral (m - n)) 1 ys)
+    | otherwise = Laurent m (Poly.scale (fromIntegral (n - m)) 1 xs + ys)
+  negate (Laurent n xs) = Laurent n (negate xs)
+  abs    = id
+  signum = const 1
+  fromInteger n = Laurent 0 (fromInteger n)
+  {-# INLINE (+) #-}
+  {-# INLINE negate #-}
+  {-# INLINE fromInteger #-}
+  {-# INLINE (*) #-}
+
+quotRemLaurent :: (Eq a, Integral a, G.Vector v a) => Laurent v a -> Laurent v a -> (Laurent v a, Laurent v a)
```

Comment:
As mentioned in #23, it is impossible to implement division with remainder for Laurent polynomials. E. g., imagine `(q, r) = quotRemLaurent (X + 1) (X - 1)`. Is it `(q, r) = (1, 2)` or `(q, r) = (1 + 2 * X^(-1), 2 * X^(-1))` or what?

However, one can still define `gcd` and `lcm`.